React/README.md

```diff
@@ -0,0 +1,276 @@
+
+  Плохо:
+  ```typescript
+  const MyComponent: FC<Props> = ({ ... }) => {
+    const DAYS_IN_WEEK = 7;
+
+    const calculateSomeValue = () => { ... };
+
+    return ( ... );
+  };
+  ```
+  Хорошо:
+  ```typescript
+  // Константы здесь рядом с компонентом для примера, их надо вынести в отдельные файлы.
+  const DAYS_IN_WEEK = 7;
+
+  const calculateSomeValue = () => { ... };
+
+  const MyComponent: FC<Props> = ({ ... }) => {
+    return ( ... );
+  };
+  ```
+
+<a name="2"></a>
+
+## Пропсы
+
+<a name="2.1"></a>
+
+- [2.1](#2.1) Пропсы компонента объявлять перед компонентом в том же файле. Это позволит не переключаться между файлами при чтении и беглом просмотре компонента. 
+
+<a name="2.2"></a>
+
+- [2.2](#2.2) Именовать пропсы компонента словом `Props`, передавая их в дженерик `FC`, что бы получилась запись вида:
+  ```typescript
+  const MyComponent: FC<Props> = ({ ... }) => ...
+  ```
+
+  Более подробное имя для пропсов не требуется, из контекста и так понятно, что это пропсы для текущего компонента.
+
+<a name="2.3"></a>
+
+- [2.3](#2.3) Если ваш компонент предполагает обязательное использование пропа `children`, явно объявите его в `Props` и сделайте обязательным. Если `children` необязательный, можно воспользоваться вспомогательным встроенным типом `PropsWithChildren<Props>`.
+
+<a name="2.4"></a>
+
+- [2.4](#2.4) Не экспортируйте пропсы компонента без причины. Во первых это засоряет подсказки редактора кода, во вторых неясно, эти пропсы действительно где-то используются или нет. Экспортируйте только если они используются где-то ещё. При экспорте пропсам следует дать более осмысленное имя, добавив название компонента в начало: 
+  ```typescript
+  export { type Props as MyComponentProps };
+  ```
+  Замечание: правило применимо, если вы пропсы объявляете в том же файле, что и сам компонент.
+
+<a name="2.5"></a>
+
+- [2.5](#2.5) Порядок указания пропсов должен быть следующим - все обязательные значения, все необязательные значения, все обязательные функции, все необязательные функции. Все необязательные пропсы должны иметь значение по умолчанию, за исключением функций.
+
+<a name="2.6"></a>
+
+- [2.6](#2.6) Пропсы компонента, относящиеся к обработке событий, должны именоваться по шаблону `onEventName`, где `EventName` — имя обрабатываемого события.
+
+<a name="3"></a>
+
+## SVG
+
+<a name="3.1"></a>
+
+- [3.1](#3.1) Для svg-иконок задавайте свойство `fill` в значение `currentColor`. Так можно будет менять цвет иконки из CSS.
+
+<a name="3.2"></a>
+
+- [3.2](#3.2) Оборачивайте svg-код в собственные реакт компоненты. Во многих фреймворках встроен SVGR, который позволяет подключать svg-файлы как реакт-компоненты. Проблема в том, что у такого компонента напрочь отсутствует типизация, ему можно передать любые пропсы. Поэтому вместо svg-файла стоит создать собственный реакт-компонент, который будет возвращать svg-код. И явно задавать пропсы.  
```

Comment:
данная проблема решается проще: в .d.ts фале:
```tsx
declare module '*.svg' {
  import * as React from 'react';

  export const ReactComponent: React.FunctionComponent<
    React.SVGProps<SVGSVGElement> & { title?: string }
  >;

  const src: string;
  export default src;
}
```

Reply:
хм, здорово, спасибо) Полагаю, что иногда всё таки может понадобиться svg как реакт компонент, например, что бы по специфическому пропсу описать тот или иной внешний вид. Но по идее это частный кейс который можно обработать отдельно.

Reply:
> что иногда всё таки может понадобиться svg как реакт компонент, например, что бы по специфическому пропсу описать тот или иной внешний вид.

Для изменения внешнего вида лучше просто использовать альтернативный файл .svg: например при изменении размера часто отличается и структура: убираются некоторые детали/просто меняется дизайн. Я бы сказал что в отдельную компоненту имеет смысл выносить только если нам нужно что-то по типу анимаций.

Reply:
> данная проблема решается проще: в .d.ts файле:

Можешь подсказать как это правильно сделать? Например я в проекте Next.js в корне создал `global.d.ts` файл, вставил туда этот код. Так же добавил в `tsconfig.json` в массив `include` путь к файлу. Но пока не заработало. Что-то делаю не так?

Reply:
> > данная проблема решается проще: в .d.ts файле:
> 
> Можешь подсказать как это правильно сделать? Например я в проекте Next.js в корне создал `global.d.ts` файл, вставил туда этот код. Так же добавил в `tsconfig.json` в массив `include` путь к файлу. Но пока не заработало. Что-то делаю не так?

Проблема в том, что `next-env.d.ts` ссылаясь на `/// <reference types="next/image-types/global" />` в котором `default` обозначен как `any`. Соответственно при слиянии такого определения и пользовательского any + string -> any. Не у верен что это поведение можно как-то изменить.

```ts
declare module '*.svg' {
  /**
   * Use `any` to avoid conflicts with
   * `@svgr/webpack` plugin or
   * `babel-plugin-inline-react-svg` plugin.
   */
  const content: any

  export default content
}

```

P.S.: импорт `ReactComponent` работает как надо т.к. next не задаёт ему тип:
![image](https://user-images.githubusercontent.com/104826556/222405304-96dc7acf-5577-4086-84d1-664c09b1379e.png)

Reply:
Думал как описать это в бест практис... теперь получается, что не так уж и проще) Придётся при импорте svg использовать переименование `import { ReactComponent as MyIcon } from './MyIcon.svg';`. И возможно кому-то будет удобней оборачивать в реакт-компоненты. 

Можно так: описать проблему и предложить 2 решения: 1) оборачивать в реакт-компоненты и указывать пропсы, 2) типизировать svg-файлы, показав как это делается.

Reply:
можно типизировать глобально и экспортировать из index:

![image](https://user-images.githubusercontent.com/104826556/222426121-76842cb2-04ac-4a76-946d-616d03902716.png)

Так и писать меньше, и импорт комфортнее)

Reply:
Поправил. Получилось как небольшое руководство по решению проблемы. Пока останется здесь, но после бест практисов я планировал заняться созданием репо по типу "база знаний фронтенда" (подробности потом). Это руководство можно будет вынести туда, а здесь просто оставить ссылку.